src/analysis/souffle/fact_test_helper.dl

```diff
@@ -3,15 +3,27 @@
 
 #include "taint.dl"
 
-.type TestName <: symbol
-
-// If a test has something in this relation, then it has passed. If not, then it has failed. Each
-// test in arcs_fact_tests must provide a rule that populates this relation indicating the
-// expectations that must be met for the test to pass. We check for whether this relation has
-// non-zero size in the test driver to decide whether to pass or fail the test. Although the test
-// driver does not care about the contents of this relation, we use the test name as the parameter
-// so printing the relation provides some useful information.
-.decl testPasses(testName: TestName)
-.output testPasses(IO=stdout, delimiter=",")
+.type TestAspectName <: symbol
+
+// Any tests in this rule have passed. Should usually be populated by TEST_RULE rather than
+// directly used.
+.decl testPasses(testAspectName: TestAspectName)
+
+// If a the testFails relation has any contents, then the test has failed. If not, it has passed.
+// Tests should create rules indicating what should not be true of the test for it to pass. Should
+// usually be populated by TEST_RULE rather than directly used.
+.decl testFails(testAspectName: TestAspectName)
+.output testFails(IO=stdout, delimiter=",")
+
+.decl allTests(testAspectName: TestAspectName)
+.output allTests(IO=stdout, delimiter=",")
+
+allTests(testAspectName) :- testPasses(testAspectName).
+allTests(testAspectName) :- testFails(testAspectName).
+
+#define TEST_RULE(test_aspect_name, rule) \
+  testPasses(test_aspect_name) :- rule. \
+  testFails(test_aspect_name) :- !(rule)
+
```

Comment:
How about tweaking the macro a little bit. 

```suggestion
#define TEST_RULE(test_aspect_name, rule) \
  isTest(test_aspsect_name). \
  testPasses(test_aspect_name)
```

Then you can use the macro as follows:
```
TEST_RULE("should_not_be_tagged_with_trusted") :-  !mayHaveTag("R.P2.bar", "trusted").
```

instead of 
```
TEST_RULE("should_not_be_tagged_with_trusted", !mayHaveTag("R.P2.bar", "trusted")).
```

Now, `isTest` is to true for all the declared tests. Then we can go add the following rule:
```
testFails(x) :- isTest(x), !testPasses(x).
```
Further, we can simply do the following in `fact_rule_checker` to detect test failures:
  - `isTest` is not empty.
  - `testFails` is empty.

Reply:
This is excellent, I like it a lot. This also allows `TEST_RULE` rules to use `,` as `AND`. Going to implement this now.

Reply:
Done, PTAL.